File: shared/PocketCore/Sources/PocketCore/Calculator/CalculatorEngine.swift
// CalculatorEngine — pure-Swift expression evaluator using shunting-yard.
// Supports: + - × ÷ % ( ) sin cos tan log ln √ x² x^y π e
// Returns .number on success, .error otherwise.
import Foundation

public enum CalcResult: Equatable {
    case number(Double)
    case error(String)
}

public enum CalculatorEngine {
    /// Evaluate the given infix expression string.
    /// Whitespace is ignored. Operators accepted: + - * / × ÷ % ^
    /// Functions: sin cos tan asin acos atan log ln sqrt
    /// Constants: pi π e
    /// Unary minus accepted at start or after another operator/(.
    public static func evaluate(_ expression: String) -> CalcResult {
        do {
            let tokens = try tokenize(expression)
            let rpn = try toRPN(tokens)
            let value = try evalRPN(rpn)
            if value.isNaN { return .error("Not a number") }
            if value.isInfinite { return .error("Overflow") }
            return .number(value)
        } catch let e as CalcError {
            return .error(e.message)
        } catch {
            return .error("Invalid expression")
        }
    }

    // MARK: - Tokenizer

    private enum Token: Equatable {
        case number(Double)
        case op(String)         // + - * / ^ %
        case fn(String)         // sin cos … sqrt
        case lparen, rparen
        case unaryMinus
    }

    private struct CalcError: Error { let message: String }

    private static func tokenize(_ raw: String) throws -> [Token] {
        var tokens: [Token] = []
        let normalized = raw
            .replacingOccurrences(of: "×", with: "*")
            .replacingOccurrences(of: "÷", with: "/")
            .replacingOccurrences(of: "−", with: "-")
            .replacingOccurrences(of: "√", with: "sqrt")
            .replacingOccurrences(of: "π", with: "pi")
        let chars = Array(normalized)
        var i = 0
        while i < chars.count {
            let c = chars[i]
            if c.isWhitespace { i += 1; continue }
            if c.isNumber || c == "." {
                var s = ""
                while i < chars.count, chars[i].isNumber || chars[i] == "." {
                    s.append(chars[i]); i += 1
                }
                guard let d = Double(s) else { throw CalcError(message: "Bad number \(s)") }
                tokens.append(.number(d))
                continue
            }
            if c.isLetter {
                var s = ""
                while i < chars.count, chars[i].isLetter {
                    s.append(chars[i]); i += 1
                }
                let lower = s.lowercased()
                switch lower {
                case "pi":  tokens.append(.number(.pi))
                case "e":   tokens.append(.number(M_E))
                case "sin", "cos", "tan", "asin", "acos", "atan", "log", "ln", "sqrt":
                    tokens.append(.fn(lower))
                default: throw CalcError(message: "Unknown identifier \(lower)")
                }
                continue
            }
            switch c {
            case "(":
                tokens.append(.lparen)
            case ")":
                tokens.append(.rparen)
            case "+":
                tokens.append(.op("+"))
            case "-":
                if isUnaryContext(prev: tokens.last) {
                    tokens.append(.unaryMinus)
                } else {
                    tokens.append(.op("-"))
                }
            case "*":
                tokens.append(.op("*"))
            case "/":
                tokens.append(.op("/"))
            case "^":
                tokens.append(.op("^"))
            case "%":
                tokens.append(.op("%"))
            default:
                throw CalcError(message: "Unexpected '\(c)'")
            }
            i += 1
        }
        return tokens
    }

    private static func isUnaryContext(prev: Token?) -> Bool {
        guard let prev else { return true }
        switch prev {
        case .op, .lparen, .unaryMinus, .fn: return true
        default: return false
        }
    }

    // MARK: - Shunting yard

    private static func precedence(_ op: String) -> Int {
        switch op {
        case "+", "-": return 1
        case "*", "/", "%": return 2
        case "^": return 3
        default: return 0
        }
    }

    private static func isRightAssoc(_ op: String) -> Bool { op == "^" }

    private static func toRPN(_ tokens: [Token]) throws -> [Token] {
        var out: [Token] = []
        var stack: [Token] = []
        for tok in tokens {
            switch tok {
            case .number:
                out.append(tok)
            case .fn:
                stack.append(tok)
            case .unaryMinus:
                stack.append(tok)
            case .op(let o):
                while let top = stack.last {
                    if case .op(let topOp) = top {
                        let p = precedence(o), tp = precedence(topOp)
                        if (tp > p) || (tp == p && !isRightAssoc(o)) {
                            out.append(stack.removeLast())
                            continue
                        }
                        break
                    } else if case .unaryMinus = top {
                        // unary minus has very high precedence, drain it
                        out.append(stack.removeLast())
                    } else if case .fn = top {
                        out.append(stack.removeLast())
                    } else {
                        break
                    }
                }
                stack.append(tok)
            case .lparen:
                stack.append(tok)
            case .rparen:
                var found = false
                while let top = stack.last {
                    if case .lparen = top {
                        stack.removeLast()
                        found = true
                        break
                    }
                    out.append(stack.removeLast())
                }
                if !found { throw CalcError(message: "Mismatched parens") }
                if let top = stack.last, case .fn = top {
                    out.append(stack.removeLast())
                }
            }
        }
        while let top = stack.popLast() {
            if case .lparen = top { throw CalcError(message: "Mismatched parens") }
            if case .rparen = top { throw CalcError(message: "Mismatched parens") }
            out.append(top)
        }
        return out
    }

    private static func evalRPN(_ rpn: [Token]) throws -> Double {
        var st: [Double] = []
        for tok in rpn {
            switch tok {
            case .number(let d):
                st.append(d)
            case .unaryMinus:
                guard let a = st.popLast() else { throw CalcError(message: "Bad unary") }
                st.append(-a)
            case .op(let o):
                guard let b = st.popLast(), let a = st.popLast() else {
                    throw CalcError(message: "Bad operands for \(o)")
                }
                switch o {
                case "+": st.append(a + b)
                case "-": st.append(a - b)
                case "*": st.append(a * b)
                case "/":
                    if b == 0 { throw CalcError(message: "Divide by zero") }
                    st.append(a / b)
                case "%":
                    if b == 0 { throw CalcError(message: "Divide by zero") }
                    st.append(a.truncatingRemainder(dividingBy: b))
                case "^": st.append(pow(a, b))
                default: throw CalcError(message: "Unknown op \(o)")
                }
            case .fn(let f):
                guard let a = st.popLast() else { throw CalcError(message: "Bad arg for \(f)") }
                switch f {
                case "sin":  st.append(sin(a))
                case "cos":  st.append(cos(a))
                case "tan":  st.append(tan(a))
                case "asin": st.append(asin(a))
                case "acos": st.append(acos(a))
                case "atan": st.append(atan(a))
                case "log":  st.append(log10(a))
                case "ln":   st.append(log(a))
                case "sqrt":
                    if a < 0 { throw CalcError(message: "Sqrt of negative") }
                    st.append(sqrt(a))
                default: throw CalcError(message: "Unknown fn \(f)")
                }
            case .lparen, .rparen:
                throw CalcError(message: "Stray paren")
            }
        }
        guard st.count == 1, let r = st.first else { throw CalcError(message: "Invalid expression") }
        return r
    }
}

File: apps/ClockApp/android/core/src/main/java/com/americangroupllc/pocket/core/calculator/CalculatorEngine.kt
package com.americangroupllc.pocket.core.calculator

import kotlin.math.*

sealed class CalcResult {
    data class Number(val value: Double) : CalcResult()
    data class Error(val message: String) : CalcResult()
}

/**
 * Pure-Kotlin shunting-yard expression evaluator.
 * Supports: + - * / × ÷ % ^ ( ) sin cos tan asin acos atan log ln sqrt √ pi π e
 */
object CalculatorEngine {

    fun evaluate(expression: String): CalcResult {
        return try {
            val tokens = tokenize(expression)
            val rpn = toRPN(tokens)
            val v = evalRPN(rpn)
            when {
                v.isNaN() -> CalcResult.Error("Not a number")
                v.isInfinite() -> CalcResult.Error("Overflow")
                else -> CalcResult.Number(v)
            }
        } catch (e: CalcException) {
            CalcResult.Error(e.message ?: "Invalid expression")
        }
    }

    private class CalcException(message: String) : RuntimeException(message)

    private sealed class Token {
        data class Num(val v: Double) : Token()
        data class Op(val s: String) : Token()
        data class Fn(val s: String) : Token()
        object L : Token()
        object R : Token()
        object UnaryMinus : Token()
    }

    private fun tokenize(raw: String): List<Token> {
        val s = raw
            .replace("×", "*")
            .replace("÷", "/")
            .replace("−", "-")
            .replace("√", "sqrt")
            .replace("π", "pi")
        val out = mutableListOf<Token>()
        var i = 0
        while (i < s.length) {
            val c = s[i]
            if (c.isWhitespace()) { i++; continue }
            if (c.isDigit() || c == '.') {
                val sb = StringBuilder()
                while (i < s.length && (s[i].isDigit() || s[i] == '.')) {
                    sb.append(s[i]); i++
                }
                out.add(Token.Num(sb.toString().toDoubleOrNull() ?: throw CalcException("Bad number")))
                continue
            }
            if (c.isLetter()) {
                val sb = StringBuilder()
                while (i < s.length && s[i].isLetter()) {
                    sb.append(s[i]); i++
                }
                when (val name = sb.toString().lowercase()) {
                    "pi" -> out.add(Token.Num(PI))
                    "e"  -> out.add(Token.Num(E))
                    "sin", "cos", "tan", "asin", "acos", "atan", "log", "ln", "sqrt" ->
                        out.add(Token.Fn(name))
                    else -> throw CalcException("Unknown identifier $name")
                }
                continue
            }
            when (c) {
                '(' -> out.add(Token.L)
                ')' -> out.add(Token.R)
                '+' -> out.add(Token.Op("+"))
                '-' -> if (isUnaryContext(out.lastOrNull())) out.add(Token.UnaryMinus) else out.add(Token.Op("-"))
                '*' -> out.add(Token.Op("*"))
                '/' -> out.add(Token.Op("/"))
                '^' -> out.add(Token.Op("^"))
                '%' -> out.add(Token.Op("%"))
                else -> throw CalcException("Unexpected '$c'")
            }
            i++
        }
        return out
    }

    private fun isUnaryContext(prev: Token?): Boolean {
        if (prev == null) return true
        return prev is Token.Op || prev is Token.L || prev is Token.UnaryMinus || prev is Token.Fn
    }

    private fun precedence(op: String) = when (op) {
        "+", "-" -> 1
        "*", "/", "%" -> 2
        "^" -> 3
        else -> 0
    }

    private fun isRightAssoc(op: String) = op == "^"

    private fun toRPN(tokens: List<Token>): List<Token> {
        val out = mutableListOf<Token>()
        val stack = ArrayDeque<Token>()
        for (tok in tokens) {
            when (tok) {
                is Token.Num -> out.add(tok)
                is Token.Fn -> stack.addLast(tok)
                is Token.UnaryMinus -> stack.addLast(tok)
                is Token.Op -> {
                    while (true) {
                        val top = stack.lastOrNull() ?: break
                        when (top) {
                            is Token.Op -> {
                                val tp = precedence(top.s)
                                val p = precedence(tok.s)
                                if (tp > p || (tp == p && !isRightAssoc(tok.s))) {
                                    out.add(stack.removeLast())
                                    continue
                                } else break
                            }
                            is Token.UnaryMinus, is Token.Fn -> out.add(stack.removeLast())
                            else -> break
                        }
                    }
                    stack.addLast(tok)
                }
                Token.L -> stack.addLast(tok)
                Token.R -> {
                    var found = false
                    while (stack.isNotEmpty()) {
                        val top = stack.removeLast()
                        if (top is Token.L) { found = true; break }
                        out.add(top)
                    }
                    if (!found) throw CalcException("Mismatched parens")
                    val maybeFn = stack.lastOrNull()
                    if (maybeFn is Token.Fn) out.add(stack.removeLast())
                }
            }
        }
        while (stack.isNotEmpty()) {
            val t = stack.removeLast()
            if (t is Token.L || t is Token.R) throw CalcException("Mismatched parens")
            out.add(t)
        }
        return out
    }

    private fun evalRPN(rpn: List<Token>): Double {
        val st = ArrayDeque<Double>()
        for (tok in rpn) {
            when (tok) {
                is Token.Num -> st.addLast(tok.v)
                is Token.UnaryMinus -> {
                    val a = st.removeLastOrNull() ?: throw CalcException("Bad unary")
                    st.addLast(-a)
                }
                is Token.Op -> {
                    val b = st.removeLastOrNull() ?: throw CalcException("Bad op")
                    val a = st.removeLastOrNull() ?: throw CalcException("Bad op")
                    val r = when (tok.s) {
                        "+" -> a + b
                        "-" -> a - b
                        "*" -> a * b
                        "/" -> { if (b == 0.0) throw CalcException("Divide by zero"); a / b }
                        "%" -> { if (b == 0.0) throw CalcException("Divide by zero"); a % b }
                        "^" -> a.pow(b)
                        else -> throw CalcException("Unknown op ${tok.s}")
                    }
                    st.addLast(r)
                }
                is Token.Fn -> {
                    val a = st.removeLastOrNull() ?: throw CalcException("Bad fn")
                    val r = when (tok.s) {
                        "sin" -> sin(a); "cos" -> cos(a); "tan" -> tan(a)
                        "asin" -> asin(a); "acos" -> acos(a); "atan" -> atan(a)
                        "log" -> log10(a); "ln" -> ln(a)
                        "sqrt" -> { if (a < 0) throw CalcException("Sqrt of negative"); sqrt(a) }
                        else -> throw CalcException("Unknown fn ${tok.s}")
                    }
                    st.addLast(r)
                }
                else -> throw CalcException("Stray paren")
            }
        }
        if (st.size != 1) throw CalcException("Invalid expression")
        return st.last()
    }
}
